cmd2/AutoCompleter.py

```diff
@@ -0,0 +1,796 @@
+                                          # action_name -> dict(sub_command -> completer)
+
+        # Start digging through the argparse structures.
+        #   _actions is the top level container of parameter definitions
+        for action in self._parser._actions:
+            # if there are choices defined, record them in the arguments dictionary
+            if action.choices is not None:
+                self._arg_choices[action.dest] = action.choices
+
+            # if the parameter is flag based, it will have option_strings
+            if action.option_strings:
+                # record each option flag
+                for option in action.option_strings:
+                    self._flags.append(option)
+                    self._flag_to_action[option] = action
+                    if action.nargs == 0:
+                        self._flags_without_args.append(option)
+            else:
+                self._positional_actions.append(action)
+
+                if isinstance(action, argparse._SubParsersAction):
+                    sub_completers = {}
+                    sub_commands = []
+                    args_for_action = subcmd_args_lookup[action.dest]\
+                        if action.dest in subcmd_args_lookup.keys() else {}
+                    for subcmd in action.choices:
+                        (subcmd_args, subcmd_lookup) = args_for_action[subcmd] if \
+                            subcmd in args_for_action.keys() else \
+                            (arg_choices, subcmd_args_lookup) if forward_arg_choices else ({}, {})
+                        subcmd_start = token_start_index + len(self._positional_actions)
+                        sub_completers[subcmd] = AutoCompleter(action.choices[subcmd], subcmd_start,
+                                                               arg_choices=subcmd_args,
+                                                               subcmd_args_lookup=subcmd_lookup)
+                        sub_commands.append(subcmd)
+                    self._positional_completers[action.dest] = sub_completers
+                    self._arg_choices[action.dest] = sub_commands
+
+    def complete_command(self, tokens: List[str], text: str, line: str, begidx: int, endidx: int) -> List[str]:
+        """Complete the command using the argparse metadata and provided argument dictionary"""
+        # Count which positional argument index we're at now. Loop through all tokens on the command line so far
+        # Skip any flags or flag parameter tokens
+        next_pos_arg_index = 0
+
+        pos_arg = AutoCompleter._ArgumentState()
+        pos_action = None
+
+        flag_arg = AutoCompleter._ArgumentState()
+        flag_action = None
+
+        matched_flags = []
+        current_is_positional = False
+        consumed_arg_values = {}  # dict(arg_name -> [values, ...])
+
+        def consume_flag_argument() -> None:
+            """Consuming token as a flag argument"""
```

Comment:
While I believe that this function and the next one do work properly within the context of your code, it feels slightly awkward that they are using the ``token`` local variable which isn't passed in to them, but exists within the calling function's scope.

Is there a good reason for doing it this way?

I guess it is OK.  But consider possibly refactoring.  It just took me a little longer to read the code and figure out what is going on.  Initially I was like "WTF ``token`` is an undefined variable".

Reply:
These functions need a lot of the variables from the parent function. I would need to pass a lot of state arguments to a nested function that's clearly isolated within the scope of this function. It is a bit weird that it lets you access variables declared after this point. Python scoping is a bit weird at times.

Reply:
Maybe just put a brief comment?